probably/pgcl/parser.py

```diff
@@ -313,30 +388,94 @@ def _parse_instrs(t: Tree) -> List[Instr]:
     return [_parse_instr(_as_tree(t)) for t in t.children]
 
 
-def _parse_program(t: Tree) -> Program:
+def _parse_queries(t: Tree) -> List[Instr]:
+    assert t.data == "queries"
+    return [_parse_query(_as_tree(t)) for t in t.children]
+
+
+def _parse_query(t: Tree):
+    if t.data == 'expectation':
+        return ExpectationInstr(_parse_expr(_child_tree(t, 0)))
+    elif t.data == 'prquery':
+        return ProbabilityQueryInstr(_parse_expr(_child_tree(t, 0)))
+    elif t.data == 'print':
+        return PrintInstr()
+    elif t.data == 'optimize':
+        mode = _parse_var(_child_tree(t, 2))
+        if mode == "MAX":
+            opt_type = OptimizationType.MAXIMIZE
+        elif mode == "MIN":
+            opt_type = OptimizationType.MINIMIZE
+        else:
+            raise SyntaxError(f"The Optimization can either be 'MAX' or 'MIN', but not {mode}")
+        parameter = _parse_var(_child_tree(t, 1))
+        if parameter not in parameters:
+            raise SyntaxError(
+                f"In Optimization queries, the variable can only be a program parameter, was {parameter}.")
+        return OptimizationQuery(_parse_expr(_child_tree(t, 0)), parameter, opt_type)
+    elif t.data == "plot":
+        if len(t.children) == 3:
+            lit = _parse_literal(_child_tree(t, 2))
+            if isinstance(lit, BoolLitExpr):
+                raise SyntaxError("Plot instructions cannot handle boolean literals as arguments")
+            if t.children[2].data == 'real' or t.children[2].data == 'infinity':
+                return PlotInstr(VarExpr(_parse_var(_child_tree(t, 0))),
+                                 VarExpr(_parse_var(_child_tree(t, 1))),
+                                 prob=lit)
+            else:
+                return PlotInstr(VarExpr(_parse_var(_child_tree(t, 0))),
+                                 VarExpr(_parse_var(_child_tree(t, 1))),
+                                 term_count=lit)
+        elif len(t.children) == 2:
+            if t.children[1].data == 'real':
+                lit = _parse_literal(_child_tree(t, 1))
+                return PlotInstr(VarExpr(_parse_var(_child_tree(t, 0))),
+                                 prob=lit)
+            elif t.children[1].data == 'nat':
+                lit = _parse_literal(_child_tree(t, 1))
+                return PlotInstr(VarExpr(_parse_var(_child_tree(t, 0))),
+                                 term_count=lit)
+            elif t.children[1].data == 'infinity':
+                lit = _parse_literal(_child_tree(t, 1))
+                return PlotInstr(VarExpr(_parse_var(_child_tree(t, 0))),
+                                 prob=lit)
+            elif t.children[1].data == 'true' or t.children[1].data == 'false':
+                raise SyntaxError("Plot instruction does not support boolean operators")
```

Comment:
Again, belongs in the type checker.

Reply:
Is this actually resolved?

Reply:
Oops

Reply:
I think there is actually a bug in the parsing of plot instructions. If we write `!Plot[x, true]` in pGCL, the `true` is interpreted as a `VarExpr` with value `true` instead of as an (illegal) boolean literal.

Which also raises the question, shouldn't is be forbidden to have variables etc. that are called "true" or "false"?

Reply:
You're right! I guess the nicest way to solve this would be to parse full expressions in `!Plot` and then have a check somewhere that it's actually a variable. That way, we can give proper error messages.

I had hoped that the parser would reject any keywords as variables, e.g. `true := whatever` should give an error. If that's not the case, another bug! I'd solve that by manually rejecting all keywords such as `true` and `false` in the `_parse_var` function: https://github.com/Philipp15b/probably/blob/6158b3758fe62c43005e807f7e9f47cf9cf6c625/probably/pgcl/parser.py#L183-L186

Reply:
> I had hoped that the parser would reject any keywords as variables, e.g. true := whatever should give an error. If that's not the case, another bug! 

Fixed, at least for `true` and `false` (are there even more?).

> I guess the nicest way to solve this would be to parse full expressions in !Plot and then have a check somewhere that it's actually a variable.

I feel like this is just treating the symptoms (i.e., retrospectively fixing that the tree generated by lark contains a var node where a literal should be) instead of fighting the actual cause (i.e., making sure that there isn't even a var node in the first place). Wouldn't it be more elegant to somehow change the lark grammar such that `true` and `false` are never recognized as variable names?

For example, commit 12167a50e56ab6c1d88b5f9a57e930df6a848ac4 does this. This fixes both of these problems, it just leads to some very cryptic error message when a variable has an illegal name.

Reply:
> If we do it like you propose, this check needs to occur after the parsing is finished, or we need a global list in the parser that contains all variables in the current program (similar to the parameter dict)

I don't understand that. Just check whether a variable name is a keyword and throw an error if that's the case. I think that's the easiest way to do this. As an additional safety measure, the constructor of `VarExpr` could `assert` that the name is not a keyword.

> it just leads to some very cryptic error messages when a variable has an illegal name

This is exactly why we should be liberal in what we parse and then provide useful error messages afterwards.